File: 729. My Calendar I/solution.py
class MyCalendar:

    def __init__(self):
        self.st = set()

    def book(self, start: int, end: int) -> bool:
        for s, e in self.st:
            if not (end <= s or start >= e):  # Check for overlap
                return False
        
        self.st.add((start, end))  # Add the booking
        return True

File: 729. My Calendar I/solution.cpp
class MyCalendar {
public:

    set<pair<int,int>>st;
    
    MyCalendar() {
        
    }
    
    bool book(int start, int end) {
       
        auto it  = st.lower_bound({start,end});
        
        if(it!=st.end() && !(end<=it->first) || it!=st.begin() && !(prev(it)->second<=start))
            return false;
        
        
        st.insert({start,end});
        return true;
        
    }
};